pkg/agent/plugin/nodeattestor/tpmdevid/tpmutil/session.go

```diff
@@ -0,0 +1,279 @@
+// workflow.
+type Session struct {
+	devID *SigningKey
+	ak    *SigningKey
+	ek    *tpm2tools.Key
+
+	akPub []byte
+
+	rwc io.ReadWriteCloser
+	log hclog.Logger
+}
+
+type SessionConfig struct {
+	DevicePath string
+
+	DevIDPriv []byte
+	DevIDPub  []byte
+
+	Log hclog.Logger
+}
+
+// NewSession opens a connection to a TPM and configures it to be used for
+// node attestation.
+func NewSession(scfg *SessionConfig) (*Session, error) {
+	// Open TPM connection
+	tpm, err := open(scfg.DevicePath, scfg.Log)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open TPM at %q: %w", scfg.DevicePath, err)
+	}
+
+	// Close session in case of error
+	defer func() {
+		if err != nil {
+			tpm.Close()
+		}
+	}()
+
+	// Load DevID
+	tpm.devID, err = tpm.loadKey(scfg.DevIDPub, scfg.DevIDPriv)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load DevID: %w", err)
+	}
+
+	// Create Attestation Key
+	akPriv, akPub, err := tpm.createAttestationKey()
+	if err != nil {
+		return nil, fmt.Errorf("cannot create attestation key: %w", err)
+	}
+	tpm.akPub = akPub
+
+	// Load Attestation Key
+	tpm.ak, err = tpm.loadKey(akPub, akPriv)
+	if err != nil {
+		return nil, fmt.Errorf("cannot load attestation key: %w", err)
+	}
+
+	// Regenerate Endorsement Key using the default RSA template
+	tpm.ek, err = tpm2tools.EndorsementKeyRSA(tpm.rwc)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create endorsement key: %w", err)
+	}
+
+	return tpm, nil
+}
+
+// Close unloads TPM loaded objects and close the connection to the TPM.
+func (c *Session) Close() {
+	if c.devID != nil {
+		err := c.devID.Close()
+		if err != nil {
+			c.log.Warn(fmt.Sprintf("Failed to close DevID handle: %v", err))
+		}
+	}
+
+	if c.ak != nil {
+		err := c.ak.Close()
+		if err != nil {
+			c.log.Warn(fmt.Sprintf("Failed to close attestation key handle: %v", err))
+		}
+	}
+
+	if c.ek != nil {
+		c.ek.Close()
+	}
+
+	if c.rwc != nil {
+		// EmulatorReadWriteCloser type does not need to be closed. It closes
+		// the connection after each Read() call. Closing it again results in
+		// an error.
+		_, ok := c.rwc.(*tpmutil.EmulatorReadWriteCloser)
+		if ok {
+			return
+		}
+
+		err := c.rwc.Close()
+		if err != nil {
+			c.log.Warn(fmt.Sprintf("Failed to close TPM: %v", err))
+		}
+	}
+}
+
+// SolveDevIDChallenge request the TPM to sign the provided nonce using the loaded
+// DevID credentials.
+func (c *Session) SolveDevIDChallenge(nonce []byte) ([]byte, error) {
+	return c.devID.Sign(nonce)
+}
+
+// SolveCredActivationChallenge runs credential activation on the TPM. It proves
+// that the attestation key resides on the same TPM as the endorsement key.
+func (c *Session) SolveCredActivationChallenge(credentialBlob, secret []byte) ([]byte, error) {
+	hSession, err := c.createPolicySession()
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := tpm2.ActivateCredentialUsingAuth(
+		c.rwc,
+		[]tpm2.AuthCommand{
+			{Session: tpm2.HandlePasswordSession},
```

Comment:
Yes, TPMs can be optionally configured to use passwords/authorizations for the generated keys and also for each of the key hierarchies (owner, endorsement, etc). Currently, we are assuming empty passwords for all of the entities and relying on the root permissions needed to access the TPM.

I think it would be good to support this.
- Regarding hierarchy passwords, my understanding is that some OS can change the TPM hierarchies passwords and set one different that the default.
- Regarding passwords for keys (SRK and LDevIDs), that will depend on how the LDevID keys are generated. If the provisioning tool used allows the user to set a password or not.

I am happy to add password support. The most flexible scenario would include configurables for:

- Endorsement hierarchy password
- Owner hierarchy password
- Owner primary key password: Storage Root Key
- Owner ordinary keys password: DevID key

The plugin can use the given password if the configurable is set or the empty password if not set for each of the configurables.